Issue513 rect grid test improvements

draenog opened the pull request:
# Improvements to RectGrid parallelisation
## Fixes #513 

The pull requests implements the following workload in RectGrid_test:
1. Runs test file in on 1 MPI rank with known data.
2. Reads the same file in parallel and compare that the data are the same.
3. Writes the same file in parallel so it can be compared with the file written serially.

The resulting field is still different  as currently `RectGridIO::getModelState` does not read  `sst` and `sss` components
so they are missing form the resulting file.

Additionally the pull requests:
1. Tests local dimensions in parallel case
2. Introduce tests for ModelArray::Type::Z dimensions
3. Fixes reading of 3D data in paralle mode
4. Fix setting dimensions in parallel case
5. Tests local dimensions in parallel case 
6. Introduce tests for ModelArray::Type::Z dimensions

draenog commented:
@timspainNERSC Could you maybe comment on this:

> The resulting field is still different as currently RectGridIO::getModelState does not read `sst` and `sss` components
so they are missing form the resulting file.

Is it on purpose or should RectGridIO::getModelState read `ssst` and `sss`data?

timspainNERSC commented:
> @timspainNERSC Could you maybe comment on this:
> 
> > The resulting field is still different as currently RectGridIO::getModelState does not read `sst` and `sss` components
> > so they are missing form the resulting file.
> 
> Is it on purpose or should RectGridIO::getModelState read `ssst` and `sss`data?

When I wrote RectGridIO, we hadn't looked at the slab ocean, which is what uses `sss` and `sst`. These were added to ParaGridIO when I added the slab ocean, but I never went back and added them to RectGridIO. Since RectGrid is not intended to be used in the production model (it doesn't support the dynamics at all), I'm not planning on going back and adding support. Unless told otherwise, of course.

draenog commented:
> > @timspainNERSC Could you maybe comment on this:
> > > The resulting field is still different as currently RectGridIO::getModelState does not read `sst` and `sss` components
> > > so they are missing form the resulting file.
> > 
> > 
> > Is it on purpose or should RectGridIO::getModelState read `ssst` and `sss`data?
> 
> When I wrote RectGridIO, we hadn't looked at the slab ocean, which is what uses `sss` and `sst`. These were added to ParaGridIO when I added the slab ocean, but I never went back and added them to RectGridIO. Since RectGrid is not intended to be used in the production model (it doesn't support the dynamics at all), I'm not planning on going back and adding support. Unless told otherwise, of course.

In this case I wonder if we could  be worth removing these two field from ModelState in `RectGrid_test.cpp`. That way is the state is written, read and written again it should produce exactly the same file which is currently not the case.

MarionBWeinzierl commented on core/test/RectGrid_test.cpp:
```diff
@@ -104,47 +105,95 @@ TEST_CASE("Write and read a ModelState-based RectGrid restart file")
     REQUIRE(ms.data.at("hice")(targetX, targetY) < 2);
-    REQUIRE(ms.data.at("hice")(targetX, targetY) == 1.0703);
+#ifdef USE_MPI
+    REQUIRE(ms.data.at("hice")(targetX, targetY) == 1.0201 + metadataIn.localCornerY * 0.01 + metadataIn.localCornerX * 0.0001);
```
Where do these numbers come from?

draenog replied:
Good catch. They should be `xFactor` and `yFactor. I will fix it.

MarionBWeinzierl reviewed (approved):
This looks generally good to me. I just have some questions to help me better understand the changes.

MarionBWeinzierl reviewed (commented):
This looks good to me. But could you just briefly clarify the relationship between nDims possibly being larger then 2, and the partition file and bounding box etc only having x and y dimensions?

draenog commented on core/test/RectGrid_test.cpp:
```diff
@@ -39,8 +40,8 @@ TEST_CASE("Write and read a ModelState-based RectGrid restart file")
     // Fill in the data. It is not real data.
-    size_t nx = 25;
-    size_t ny = 15;
+    size_t nx = 5;
```
Yes, when debugging I often look at output of `ncdump`. It is easier for me to comprehend it with smaller dimensions  and the larger sizes do not bring anything here in my opinion. Size  5x7 is small enough to be easy to look at, and large enough to provide 3 blocks with non flat topology and not equal sizes in decomposition.

draenog commented on core/test/RectGrid_test.cpp:
```diff
@@ -104,47 +105,95 @@ TEST_CASE("Write and read a ModelState-based RectGrid restart file")
-    metadata.localExtentX = nx;
-    metadata.localExtentY = ny;
-#endif
+    int colour = MPI_UNDEFINED, key = 0;
```
It is a term used by MPI standard when defining `MPI_Comm_split` API:

> This function partitions the group associated with comm into disjoint subgroups, one for each value of color. Each subgroup contains all processes of the same color. 

There is a comment above that this part of code creates an MPI subcommunicator which contains only the first rank.

draenog commented:
> This looks good to me. But could you just briefly clarify the relationship between nDims possibly being larger then 2, and the partition file and bounding box etc only having x and y dimensions?

This comes from the design of parallelization that the domain only in x and y directions. I am not sure if the test is a good place to document it.

MarionBWeinzierl commented:
> > > @timspainNERSC Could you maybe comment on this:
> > > > The resulting field is still different as currently RectGridIO::getModelState does not read `sst` and `sss` components
> > > > so they are missing form the resulting file.
> > > 
> > > 
> > > Is it on purpose or should RectGridIO::getModelState read `ssst` and `sss`data?
> > 
> > 
> > When I wrote RectGridIO, we hadn't looked at the slab ocean, which is what uses `sss` and `sst`. These were added to ParaGridIO when I added the slab ocean, but I never went back and added them to RectGridIO. Since RectGrid is not intended to be used in the production model (it doesn't support the dynamics at all), I'm not planning on going back and adding support. Unless told otherwise, of course.
> 
> In this case I wonder if it would be worth removing these two field from ModelState in `RectGrid_test.cpp`. That way if the state is written, read and written again it should produce exactly the same file which is currently not the case.

@timspainNERSC , would that be ok?

timspainNERSC commented:
> > > > @timspainNERSC Could you maybe comment on this:
> > > > > The resulting field is still different as currently RectGridIO::getModelState does not read `sst` and `sss` components
> > > > > so they are missing form the resulting file.
> > > > 
> > > > 
> > > > Is it on purpose or should RectGridIO::getModelState read `ssst` and `sss`data?
> > > 
> > > 
> > > When I wrote RectGridIO, we hadn't looked at the slab ocean, which is what uses `sss` and `sst`. These were added to ParaGridIO when I added the slab ocean, but I never went back and added them to RectGridIO. Since RectGrid is not intended to be used in the production model (it doesn't support the dynamics at all), I'm not planning on going back and adding support. Unless told otherwise, of course.
> > 
> > 
> > In this case I wonder if it would be worth removing these two field from ModelState in `RectGrid_test.cpp`. That way if the state is written, read and written again it should produce exactly the same file which is currently not the case.
> 
> @timspainNERSC , would that be ok?

Yes, that sounds fine.

draenog commented:
I've just pushed  the updated branch. The changes:

1. Rebases on develop commit 1423515cd293a978de5d050dfafadc91756e50e0 to fix CI tests on Mac OS.
2. Remove sss and sst fields from ModelState as they are not read by ParaGridIO
3. Remove hardcoded offsets in test spotted by @MarionBWeinzierl

TomMelt commented on core/test/RectGrid_test.cpp:
```diff
@@ -24,13 +24,14 @@
 
 const std::string filename = "RectGrid_test.nc";
-const std::string partition_filename = "partition_metadata_1.nc";
+const std::string partition_filename = "partition_metadata_3.nc";
```
could you please send this partition file (`partition_metadata_3.nc`) to @timspainNERSC  so that he can upload it to the ftp server?

That way I can add it to the CI workflow.

My PR #507 `wget`s the required `.nc` files and will run the test as part of CI.

draenog replied:
Instead I've implemented generation of `partition_metadata_3.nc` file from its text dump during the building process.

TomMelt commented on core/test/RectGrid_test.cpp:
```diff
@@ -104,47 +103,95 @@ TEST_CASE("Write and read a ModelState-based RectGrid restart file")
     ModelState ms = gridIn.getModelState(filename);
 #endif
 
+#ifdef USE_MPI
```
could you add a couple more comments to this test file (`core/test/RectGrid_test.cpp`) please?

draenog replied:
I've added some more comments as requested.

TomMelt reviewed (approved):
Awesome thanks for this @draenog , this really improves the test and improves are likelihood of spotting bugs sooner. 

I would like a few more comments in the test, but otherwise I am really happy with this PR.

Thanks again :+1:
